### ISSUE TITLE
iocp reactor incorrectly duplicates some outgoing data

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3525]
|<img alt="seewind's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| seewind reported|
|-|-|
|Trac ID|trac#3525|
|Type|defect|
|Created|2008-11-06 16:12:23Z|

When using iocpreactor, send data will be repeated when net speed is slow.

Repair `abstract.py`:

```python
      def doWrite(self): 
            .... 
            .... 
            rc, bytes = self.writeToHandle(buff, evt) 
            if (rc == ERROR_IO_PENDING 
                or (not rc and numWrites >= self.maxWrites)): 
                self.offset += len(buff) //new line 
                break 
            else: 
            .... 
            .... 
```

I think `writeToHandle` will ensure send all buff data, so want correct offset.

Attachments:

* [Twisted-iocpreactor-callback.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/667/667910463dddf4a5cc75d990eb9c8bc57f8b05bd/80de57d4e30c49fc66cf35d321534f274da68abf.patch) (1393 bytes) - added by talljosh on 2014-05-26 04:17:25Z - 
* [bug3525testcase.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/667/667910463dddf4a5cc75d990eb9c8bc57f8b05bd/cad0b43538e64a553501fb415ff6431193952cb6.py) (3929 bytes) - added by talljosh on 2014-05-27 05:43:02Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3525 3525
type__defect defect
reporter__seewind seewind
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__core core
keywords__iocp iocp
time__1225987943000000 1225987943000000
changetime__1568870737745619 1568870737745619
version__None None
owner__None None
cc__thijs cc__talljosh
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to PenguinOfDoom|
|-|-|

Fixing description markup.  Also removing the _regular-releases_ milestone, as that is a milestone for release automation which this doesn't seem to be.  And re-assigning to the IOCP maintainer.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

Fixing description markup. Also wondering if priority shouldn't be lower.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Priority doesn't really mean anything in this issue tracker.  But if anything were going to be high priority, it would probably be data corruption bugs.

**Comment by twisted-trac** [comment]
|<img alt="talljosh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|talljosh commented|
|-|-|

I see this problem when my Twisted program tries to send a very large string, and apparently succeeds, then sends another short string. On the wire, a chunk of the large string is repeated.

When I add additional logging to Twisted, it looks as if this is what is happening:
 * While writing chunks of the large string, windows returns ERROR_IO_PENDING.
 * Before windows calls the callback to say that the write is complete, the Twisted program tries to send the shorter string.
 * This triggers FileHandle.resumeWriting(), but because the callback from the previous write hasn't been called, the buffer offset hasn't advanced, so the same chunk of data is sent again.

The attached patch fixes this problem for me. The patch is against Twisted 13.1.0, but that code hasn't changed in latest trunk.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to talljosh|
|-|-|

Thanks.  Can you please write a unit test for this as well?

**Comment by twisted-trac** [comment]
|<img alt="talljosh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|talljosh set owner to @exarkun|
|-|-|

I'm not quite sure how to unit test for this bug - so far I can only replicate the problem over a slow network connection. Any pointers? Or is it enough to just do unit tests that make sure that this change hasn't broken existing functionality?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

The unit test should fail without the patch and pass with the patch (else we have no way of knowing the bug has been fixed).  Also, if there are other important behaviors provided by the code being changed which aren't tested, tests should be added for those as well (else we have no way of knowing if they've been broken or not).

Why is a slow network connection required to reproduce the issue?  Can you use a fast network connection (ie, loopback) and a slow reader instead (call `pauseProducing` on the transport on the receiving side, for example)?  Or is the network performance really the essential factor because of some mysterious thing going on inside the Windows network stack?

**Comment by twisted-trac** [comment]
|<img alt="talljosh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|talljosh commented|
|-|-|

Using `pauseProducing` on loopback does indeed allow me to replicate. I've attached a unit test that fails with current trunk and passes once the patch is applied. The test case is a bit involved but I couldn't find a simpler way of triggering the bug.

I also don't know how to specify that the test should be skipped on other platforms. Can I leave that kind of detail with you exarkun?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban commented|
|-|-|

Here is my help for this... feel free to ignore any comment and wait for feedback from exarkun

-----

You can skip a test on platforms other than window

```
    def test_windows_support(self):
        # windows specific testing code
        pass
    if not platform.isWindows():
        test_windows_support.skip = 'Describe why the test is skipped'
```

You can also skip a whole test suite

```
class SomeTest(unittest.TestCase):
    if not platform.isWinodws():
        skip = 'Describe skip reason'

    def test_windows_stuff(self):
        pass

    def test_other_windows_stuff(self):
        pass

```

right now trial does not support skipUnless decorator... so
you can not use this:

```
    
    @unittest.skipUnless(sys.platform.startswith("win"), "requires Windows")
    def test_windows_support(self):
        # windows specific testing code
        pass

```


---------

Please integrate your test into the test suite... for example in
twisted.internet.test.test_iocp 

--------

Don't forget to add a news file for your patch
http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

and make sure the patch is submitted as a single file
-------

You might wan to make waitingForCallback, private: ex called it
_waitingForCallback

------

Your system/integration test is great, but I think that you can also write a unit test.

Your patch is just handling the case when writeToHandle returns ERROR_IO_PENDING

This is why you can start with writing a DummyIOCPReactor which inherits from the abstract.FileHandler and which implements
a writeToHandle which in some condition returns ERROR_IO_PENDING...
for example when it has received more than N bytes ... your test will then check the behavior of doWrite to make sure that it pause while IO_PENDING...

-----

Not sure what would happen if the reactor received a lot of data and it does not manage to write it... I guess that the buffer will continue to grow, so maybe this also needs some code to check for buffer size and automatically pause the producer... resume the producer after buffer is written 

-----

bug3525testcase.py starts a real reactor and uses the network, which makes the test slow and in the same time make the test
vulnerable to leaving the reactor in a different state that the one before the test.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to talljosh|
|-|-|

Seems like adiroiban provided some good review feedback.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

[#8870](https://github.com/twisted/twisted/issues/8870) is, if not a duplicate of this, a manifestation of much similar behavior.

**Comment by twisted-trac** [comment]
|<img alt="talljosh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|talljosh commented|
|-|-|

Does the attached patch (https://twistedmatrix.com/trac/attachment/ticket/3525/Twisted-iocpreactor-callback.patch) fix the behaviour observed in [#8870](https://github.com/twisted/twisted/issues/8870)?

**Comment by twisted-trac** [comment]
|<img alt="talljosh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|talljosh removed owner|
|-|-|

Reassigning to nobody, as I simply do not have to time to correctly integrate the test I provided into Twisted. I've provided a patch and a first past at tests above, perhaps someone more familiar with contributing to Twisted may be able to do whatever final steps are required for this one.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

Replying to [talljosh](#note_16):
> Does the attached patch (https://twistedmatrix.com/trac/attachment/ticket/3525/Twisted-iocpreactor-callback.patch) fix the behaviour observed in [#8870](https://github.com/twisted/twisted/issues/8870)?

No, it appears not.  I thought it would :-\.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

Now that [#8870](https://github.com/twisted/twisted/issues/8870) is fixed it would be good if someone could see if this still happens.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

[#9446](https://github.com/twisted/twisted/issues/9446) might be a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph set status to `closed`|
|-|-|

Pretty sure that this was the issue fixed in [#9446](https://github.com/twisted/twisted/issues/9446).

If anyone has evidence to the contrary, please feel free to comment and reopen.
